rfcs/0181-list-index-syntax.md

```diff
@@ -0,0 +1,290 @@
+---
+feature: list-index-syntax
+start-date: 2024-07-14
+author: rhendric
+co-authors:
+shepherd-team: infinisil
+shepherd-leader:
+related-issues: https://github.com/NixOS/nix/issues/10949, https://github.com/NixOS/rfcs/pull/137
+---
+
+# Summary
+[summary]: #summary
+
+This proposal extends the attrpath syntax to include `'[' INT ']'` components that refer to elements in lists.
+This would enable expressions such as the following:
+
+```nix
+x.[0]       # = builtins.elemAt x 0
+x.[1] or y  # = if builtins.isList x && builtins.length x > 1 then builtins.elemAt x 1 else y
+x ? [3].y   # = builtins.isList x && builtins.length x > 3 && builtins.elemAt x 3 ? y
+```
+
+# Motivation
+[motivation]: #motivation
```

Comment:
@sg-qwt Let's try to keep the discussions in threads, so they can be easily marked as resolved later

---

@sg-qwt [said](https://github.com/NixOS/rfcs/pull/181#issuecomment-2228077835):

> >  When writing Nix code, it is relatively uncommon to want to index into a list, and `builtins.elemAt` suffices.
>
> My 2cents is that I'd still prefer the simplicity of `builtins.elemAt` than introducing new syntax. The usability of nix repl is poor compared to other lisp languages. However, maybe that's something can be worked on the tooling level so that writing and interacting Nix code in repl can be the same experience as writing Nix code other places.
>
> Thus, if exploring data in a REPL is the only motivation for introducing this new syntax,  I feel like it doesn't hold a very strong argument this case.

---

A primary other use case is `builtins.{match,split}`, which returns matched groups in indexed lists. A [GitHub search](https://github.com/search?q=elemAt+lang%3ANix&type=code) finds many examples of this. Though for this use case it might be better to introduce matching group labels, so that `(builtins.match "(?<foo>.*)" "bar").foo == "bar"` would work.